File: src/models/contract.go
package models

import (
	"github.com/Acsigen/pfa-manager-api/database"
)

// Contract model, used tags to set the required items
type Contract struct {
	ID              int64
	UserID          int64
	ClientID        int64
	ContractNo      string `binding:"required"`
	StartDate       string `binding:"required"`
	EndDate         string `binding:"required"`
	Description     string
	CloudStorageUrl string
}

// Add a new contract
func (ctr *Contract) Add() error {
	// Build the query, use ? to avoid SQL injection
	query := `INSERT INTO contracts(contract_no, client_id, start_date, end_date, description, cloud_storage_url, user_id)
	VALUES (?,?,?,?,?,?,?)`

	// Prepare the query
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when the function call ends
	defer statement.Close()

	// Execute the query
	res, err := statement.Exec(ctr.ContractNo, ctr.ClientID, ctr.StartDate, ctr.EndDate, ctr.Description, ctr.CloudStorageUrl, ctr.UserID)
	if err != nil {
		return err
	}

	// Get the last insterted ID
	id, err := res.LastInsertId()

	// Set the ID of the client object so we can return it with the object in another function
	ctr.ID = id
	return err
}

// Method for updating a contract
func (ctr *Contract) Update() error {
	// Build the query, use ? to avoid SQL injection
	query := `UPDATE contracts
	SET contract_no = ?, start_date = ?, end_date = ?, description = ?, cloud_storage_url = ?
	WHERE id == ?`

	// Prepare the query
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when the function call ends
	defer statement.Close()

	// Execute the query
	_, err = statement.Exec(ctr.ContractNo, ctr.StartDate, ctr.EndDate, ctr.Description, ctr.CloudStorageUrl, ctr.ID)
	return err
}

// Function to get the list of contracts, return a list of contracts and error type
// This is not required to be a method since we don't really use the struct to insert data, we just create a list of contracts
func GetContractsList(userId int64, clientId int64) ([]Contract, error) {
	// Build the query
	query := "SELECT * FROM contracts where user_id == ? AND client_id == ?"

	// Direclty execute the query
	rows, err := database.DB.Query(query, userId, clientId)
	if err != nil {
		return nil, err
	}

	// Close the database connection when function call is done
	defer rows.Close()

	// Create the data structure for the return
	var contracts []Contract

	// Iterate over each row
	for rows.Next() {
		// Scan each row and map the items to client properties
		// The order of the arguments for the Scan function must be the same as the DB table not the model
		var contract Contract
		err := rows.Scan(&contract.ID,
			&contract.ContractNo,
			&contract.ClientID,
			&contract.StartDate,
			&contract.EndDate,
			&contract.Description,
			&contract.CloudStorageUrl,
			&contract.UserID)
		if err != nil {
			return nil, err
		}

		// append each client to the list of clients
		contracts = append(contracts, contract)
	}

	// Return the clients and no error
	return contracts, nil
}

// Function to retrieve the client with a specific ID
// This is not required to be a method since we don't really use the struct to insert data, we just retrieve a client from DB
func GetContractById(id int64) (*Contract, error) {
	// We build the query this way to avoid SQL injection
	query := "SELECT * FROM contracts WHERE id == ?"
	row := database.DB.QueryRow(query, id)

	// Scan the row and map the items to client properties
	var contract Contract
	err := row.Scan(&contract.ID,
		&contract.ContractNo,
		&contract.ClientID,
		&contract.StartDate,
		&contract.EndDate,
		&contract.Description,
		&contract.CloudStorageUrl,
		&contract.UserID)
	if err != nil {
		return nil, err
	}

	// Return the client
	return &contract, nil
}

// Method to delete a contract
func (c Contract) Delete(id int64) error {
	// Query statement
	query := "DELETE FROM contracts WHERE id == ?"
	statement, err := database.DB.Prepare(query)
	if err != nil {
		return err
	}

	// Close the statement when optimal
	defer statement.Close()

	// execute the query with the id of the contract
	_, err = statement.Exec(id)

	return err
}

File: src/models/contract.py
from pydantic import BaseModel
from fastapi import HTTPException
import sqlite3
from ..database import db

NOT_FOUND_ERR: str = "No such contract"

class Contract(BaseModel):
    id: int | None = None
    contract_no: str
    start_date: str
    end_date: str
    description: str | None = None
    cloud_storage_url: str | None = None
    client_id: int | None = 0
    user_id: int | None = 0

    def add(self, client_id, user_id: int):
        permitted_action: bool = check_permissions(
            client_id=client_id, current_user_id=user_id
        )
        if permitted_action:
            query = "INSERT INTO contracts(contract_no, start_date, end_date, description, cloud_storage_url, client_id, user_id) VALUES (?,?,?,?,?,?,?)"
            self.client_id = client_id
            self.user_id = user_id
            data: tuple = (
                self.contract_no,
                self.start_date,
                self.end_date,
                self.description,
                self.cloud_storage_url,
                self.client_id,
                self.user_id,
            )
            try:
                res: sqlite3.Cursor = db.execute_query(query=query, params=data)
                self.id = res.lastrowid
                return self
            except sqlite3.IntegrityError as e:
                raise HTTPException(500, e.args[0])

    def update(self, contract_id: int, client_id: int, user_id: int):
        permitted_action: bool = check_permissions(
            client_id=client_id, current_user_id=user_id, contract_id=contract_id
        )
        if permitted_action:
            query = "UPDATE contracts SET contract_no = ?, start_date = ?, end_date = ?, description = ?, cloud_storage_url = ? WHERE id = ? AND client_id == ? AND user_id == ?"
            self.id = contract_id
            self.client_id = client_id
            self.user_id = user_id
            data = (
                self.contract_no,
                self.start_date,
                self.end_date,
                self.description,
                self.cloud_storage_url,
                self.id,
                self.client_id,
                self.user_id,
            )
            try:
                checkout_contract: Contract = show_contract(
                    contract_id=contract_id, client_id=client_id, user_id=user_id
                )
                if type(checkout_contract) is Contract:
                    _: sqlite3.Cursor = db.execute_query(query=query, params=data)
                    return self
                else:
                    raise HTTPException(404, NOT_FOUND_ERR)
            except sqlite3.IntegrityError as e:
                raise HTTPException(500, e.args[0])


def show_contract(contract_id: int, client_id: int, user_id: int):
    permitted_action: bool = check_permissions(
        client_id=client_id, current_user_id=user_id, contract_id=contract_id
    )
    if permitted_action:
        query = (
            "SELECT * FROM contracts WHERE id == ? AND client_id == ? AND user_id == ?"
        )
        data = (contract_id, client_id, user_id)
        try:
            res: sqlite3.Cursor = db.execute_query(query=query, params=data)
            row: tuple = res.fetchone()
            if row:
                print(row)
                contract: Contract = Contract(
                    id=int(row[0]),
                    contract_no=row[1],
                    start_date=row[2],
                    end_date=row[3],
                    description=row[4],
                    cloud_storage_url=row[5],
                    client_id=row[6],
                    user_id=row[7],
                )
                return contract
            else:
                raise HTTPException(404, NOT_FOUND_ERR)
        except sqlite3.Error as e:
            raise HTTPException(500, e.args[0])


def list_user_contracts(client_id: int, user_id: int):
    permitted_action: bool = check_permissions(
        client_id=client_id, current_user_id=user_id
    )
    if permitted_action:
        contract_list: list[Contract] = []
        query = "SELECT * FROM contracts WHERE client_id == ? AND user_id == ?"
        data: tuple = (client_id, user_id)
        try:
            res: sqlite3.Cursor = db.execute_query(query=query, params=data)
            rows: list[tuple] = res.fetchall()
            for row in rows:
                contract: Contract = Contract(
                    id=int(row[0]),
                    contract_no=row[1],
                    start_date=row[2],
                    end_date=row[3],
                    description=row[4],
                    cloud_storage_url=row[5],
                    client_id=row[6],
                    user_id=row[7],
                )
                contract_list.append(contract)
            return contract_list
        except sqlite3.OperationalError as e:
            raise HTTPException(500, e.args[0])


def delete_user_contract(contract_id: int, client_id: int, user_id: int):
    permitted_action: bool = check_permissions(
        client_id=client_id, current_user_id=user_id, contract_id=contract_id
    )
    if permitted_action:
        query = (
            "DELETE FROM contracts WHERE id == ? AND client_id == ? AND user_id == ?"
        )
        data = (contract_id, client_id, user_id)
        try:
            checkout_contract = show_contract(
                contract_id=contract_id, client_id=client_id, user_id=user_id
            )
            if type(checkout_contract) is Contract:
                _: sqlite3.Cursor = db.execute_query(query=query, params=data)
                return True
            else:
                raise HTTPException(404, NOT_FOUND_ERR)
        except sqlite3.IntegrityError as e:
            raise HTTPException(500, e.args[0])


def check_permissions(
    client_id: int, current_user_id: int, contract_id: int | None = None
):
    if contract_id:
        query = """SELECT ctr.user_id
            FROM contracts AS ctr
            JOIN clients AS cli on ctr.client_id = cli.id
            WHERE ctr.id = ? AND cli.id = ?"""
        data = (contract_id, client_id)
    else:
        query = "SELECT user_id FROM clients WHERE id == ?"
        data = (client_id,)
    try:
        res: sqlite3.Cursor = db.execute_query(query=query, params=data)
        row: tuple = res.fetchone()
        if row:
            retrieved_user_id: int = int(row[0])
        else:
            raise HTTPException(
                status_code=403, detail="You are not allowed to perform this action"
            )
    except sqlite3.Error as e:
        raise HTTPException(500, e.args[0])
    if retrieved_user_id == current_user_id:
        return True
    else:
        raise HTTPException(
            status_code=403, detail="You are not allowed to perform this action"
        )
